### PR TITLE
Use pattern-matching in MetadataWriter for readability and possibly performance.

### PR DESCRIPTION
**Pull request by gafter**

### CODE REVIEW

File: src/Compilers/Core/Portable/PEWriter/MetadataWriter.cs [hunk 1]
```diff
@@ -1680,37 +1670,15 @@ internal EntityHandle GetTypeHandle(ITypeReference typeReference, bool treatRefA
...
-
-            IFieldDefinition fieldDef = definition as IFieldDefinition;
-            if (fieldDef != null)
-            {
-                return GetFieldDefinitionHandle(fieldDef);
-            }
-
-            IEventDefinition eventDef = definition as IEventDefinition;
-            if (eventDef != null)
+            return definition switch
```
Comment (gafter) [inline src/Compilers/Core/Portable/PEWriter/MetadataWriter.cs]: >definition [](start = 19, length = 10)

:)

File: src/Compilers/Core/Portable/PEWriter/MetadataWriter.cs [hunk 2]
```diff
@@ -3507,18 +3459,17 @@ private void SerializeMarshallingDescriptor(IMarshallingInformation marshallingI
...
-                    {
-                        writer.WriteSerializedString((string)marshaller);
-                    }
-                    else
-                    {
-                        writer.WriteByte(0);
+                        case ITypeReference marshallerTypeRef:
+                            this.SerializeTypeName(marshallerTypeRef, writer);
+                            break;
+                        case { }:
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/PEWriter/MetadataWriter.cs]: not: `case string s`?

**Comment by CyrusNajmabadi** [comment]
Did you use the IDE features for this?

File: src/Compilers/Core/Portable/PEWriter/MetadataWriter.cs (hunk 2 above)
Comment (gafter) [reply src/Compilers/Core/Portable/PEWriter/MetadataWriter.cs]: I was trying not to modify the semantics of the code as written.

---
In reply to: [291407933](https://github.com/dotnet/roslyn/pull/36219#discussion_r291407933) [](ancestors = 291407933)

**Comment by gafter** [comment]
For a few of them, yes.

---
In reply to: [499704024](https://github.com/dotnet/roslyn/pull/36219#issuecomment-499704024) [](ancestors = 499704024)

File: src/Compilers/Core/Portable/PEWriter/MetadataWriter.cs (hunk 2 above)
Comment (jcouv) [inline src/Compilers/Core/Portable/PEWriter/MetadataWriter.cs]: nit: I think we prefer the more specific `case object:` over `case { }:`

**Comment by gafter** [comment]
@dotnet/roslyn-compiler May I please have a second review from the compiler team for this tiny bug fix?

**Pull request merged**